File: apps/api/src/application/ports/payment-gateway.ts
import type { PaymentCard } from '@ecommerce/shared'

export interface ChargeRequest {
  readonly orderId: string
  readonly amount: number
  readonly card: PaymentCard
}

export type ChargeResult =
  | { readonly outcome: 'approved'; readonly reference: string }
  | { readonly outcome: 'declined'; readonly reason: string }

export interface PaymentGateway {
  charge(request: ChargeRequest): Promise<ChargeResult>
}

export const PAYMENT_GATEWAY = Symbol('PaymentGateway')

File: apps/api/src/ecommerce_api/application/ports/payment_gateway.py
from dataclasses import dataclass
from decimal import Decimal
from typing import Protocol

from ecommerce_api.application.ports.order_repository import Settlement
from ecommerce_api.application.schemas.order import PaymentCardInput


@dataclass(frozen=True, slots=True)
class ChargeRequest:
    order_id: str
    amount: Decimal
    card: PaymentCardInput


ChargeResult = Settlement


class PaymentGateway(Protocol):
    async def charge(self, request: ChargeRequest) -> ChargeResult: ...
